src/metabase/upload.clj

```diff
@@ -33,68 +33,102 @@
+;; - If we encounter a `float` (i.e. a non-zero fraction component), then we need to promote the column to a `float.`
+;;
+;; Columns can not have an abstract type, which has no meaning outside of inference and reconciliation.
+;; If we are left with an abstract type after having processed all the values, we need to traverse
+;; further up the graph until we reach a concrete `column-type`.
+;;
+;; For ease of reference and explicitness these corresponding values are given in the `abstract->concrete` map.
+;; One can figure out these mappings by simply looking up through the ancestors. For now, we require that it is always
+;; a direct ancestor, and lay out or graph so that it is the left-most one.
 
 (def ^:private h
   "This hierarchy defines a relationship between value types and their specializations.
   We use an [[metabase.util.ordered-hierarchy]] for its topological sorting, which simplify writing efficient and
   consistent implementations for of our type inference, parsing, and relaxation."
-  (-> (make-hierarchy)
-      (derive ::boolean-or-int ::boolean)
-      (derive ::boolean-or-int ::int)
-      (derive ::auto-incrementing-int-pk ::int)
-      (derive ::int ::float)
-      (derive ::date ::datetime)
-      (derive ::boolean ::varchar-255)
-      (derive ::offset-datetime ::varchar-255)
-      (derive ::datetime ::varchar-255)
-      (derive ::float ::varchar-255)
-      (derive ::varchar-255 ::text)))
+  (make-hierarchy
+   [::text
+    [::varchar-255
+     [::boolean ::*boolean-or-int*]
+     [::float
+      ;; A number value with a decimal separator, but a zero fractional component.
+      [::*float-or-int*
+       [::int
+        ;; A value that could be legally parsed as either a boolean OR an integer
+        ::*boolean-or-int*
+        ::auto-incrementing-int-pk]]]
+     [::datetime ::date]
+     [::offset-datetime]]]))
 
 (def ^:private abstract->concrete
-  "Not all value types correspond to database types. For those that don't, this maps to their concrete ancestor."
-  {::boolean-or-int ::boolean})
+  "Not all value types correspond to database types. For those that don't, this maps to their preferred concretion."
+  {::*boolean-or-int* ::boolean
+   ::*float-or-int*   ::float})
+
+(def ^:private allowed-promotions
+  "A mapping of which types a column can be implicitly relaxed to, based on the content of appended values.
+  If we require a relaxation which is not allow-listed here, we will reject the corresponding file."
+  {::int #{::float}})
+
+(def ^:private coercion-mapping
+  "A mapping of which value types should be coerced to the given existing type, rather than triggering promotion."
+  {::int #{::*float-or-int*}})
+
+(defn- coerce?
+  "Can values of the given type be coerced to the given existing column type, in a lossless fashion?"
+  [column-type value-type]
+  (contains? (coercion-mapping column-type) value-type))
```

Comment:
I'd like to keep the data first class, especially since I use it for a test.

Agree the name is poor. Since it's not a public var I think the bar is lower, but not quite as disposable as in a let. I'd be happy with an ugly but explicit name, like `column-type->coercible-value-types`. We can spitball in our sync.